resim/metrics/proto/metrics.proto

```diff
@@ -0,0 +1,296 @@
+}
+
+message UUIDArray {
+    repeated resim.proto.UUID array = 1;
+}
+
+message StringArray {
+    repeated string array = 1;
+}
+
+message MetricStatusArray {
+    repeated MetricStatus array = 1;
+}
+
+message Array {
+    oneof array {
+        DoubleArray       doubles    = 1;
+        TimestampArray    timestamps = 2;
+        UUIDArray         uuids      = 3;
+        StringArray       strings    = 4;
+        MetricStatusArray statuses   = 5;
+    }
+}
+
+message ArrayPerCategory {
+    map<string, Array> category_to_array = 1;
+}
+
+message MetricsData {
+    // Generic metrics data fields:
+    MetricsDataId   metrics_data_id = 1;
+    MetricsDataType data_type       = 2;
+    string          name            = 3;
+    string          unit            = 4;
+
+    // Fields for categorized data only:
+    bool            is_per_category = 5;
+    repeated string category_names  = 6;
+
+    // Fields for indexed data only:
+    bool            is_indexed      = 7;
+    MetricsDataId   index_data_id   = 8;
+    MetricsDataType index_data_type = 9;
+
+    // Data fields:
+    oneof data {
+        Array            array              = 10;
+        ArrayPerCategory array_per_category = 11;
+    }
+}
+
+// ---------------------
+// Metric representation
+// ---------------------
+
+enum MetricStatus {
+    NO_METRIC_STATUS             = 0;
+    PASSED_METRIC_STATUS         = 1;
+    FAILED_METRIC_STATUS         = 2;
+    NOT_APPLICABLE_METRIC_STATUS = 3;
+}
+
+enum MetricImportance {
+    NO_SPECIFIED_IMPORTANCE = 0;
+    ZERO_IMPORTANCE         = 1;
+    LOW_IMPORTANCE          = 2;
+    MEDIUM_IMPORTANCE       = 3;
+    HIGH_IMPORTANCE         = 4;
+    CRITICAL_IMPORTANCE     = 5;
+}
+
+enum MetricType {
+    NO_METRIC_TYPE               = 0;
+    DOUBLE_SUMMARY_METRIC_TYPE   = 1;
+    DOUBLE_OVER_TIME_METRIC_TYPE = 2;
+    LINE_PLOT_METRIC_TYPE        = 3;
+    BAR_CHART_METRIC_TYPE        = 4;
+    STATES_OVER_TIME_METRIC_TYPE = 5;
+    HISTOGRAM_METRIC_TYPE        = 6;
+    SCALAR_METRIC_TYPE           = 7;
+}
+
+message MetricId {
+    resim.proto.UUID id = 1;
+}
+
+message JobId {
+    resim.proto.UUID id = 1;
+}
+
+// All metrics  associated with a given job.
+// This is the type output by metrics scripts for validation.
+message JobMetrics {
+    JobId job_id = 1;
+
+    MetricCollection job_level_metrics = 2;
+
+    MetricStatus metrics_status = 3;
+
+    repeated MetricsData metrics_data = 4;
+
+    repeated MetricId metric_ids = 5;
+
+    repeated MetricsDataId metrics_data_ids = 6;
+}
+
+// Top-level generic metric collection within JobMetrics.
+// NB: This is separated out in order to allow metrics that are specifically
+// associated with higher-level concepts like batches, or lower-level constructs
+// like actors, which may (for example) PASS, even though the job-level metrics
+// FAIL.
+message MetricCollection {
+    repeated Metric metrics = 1;
+
+    MetricStatus metrics_status = 2;
```

Comment:
```suggestion
```

Reply:
Keen to leave this as I'd like to validate it. In other words:

- Whether a job passed can in theory be computed from the schema.
- I would like to validate that this lines up with user expectations.

Reply:
To clarify, @mikebauer's validator already does use these, which is my main reason to not yet get rid of it! So it will be accessed at some stage of the prod metrics pipeline.